torch/csrc/autograd/python_saved_variable_hooks.cpp

```diff
@@ -69,6 +70,7 @@ namespace torch { namespace autograd {
   }
 
   std::unique_ptr<SavedVariableHooks> PyDefaultSavedVariableHooks::get_hooks() {
+    std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
nit:
I think we can move this one after the if statement.
Keeping in mind the heavily used case where every SavedVariable creation will call here to check if there are default hooks. If we can avoid taking the lock every time it is better.

And here we are happy with returning nothing even if another thread is filling/deleting things.
Only the part where we actually read their content needs to be protected.